### ISSUE TITLE
remote.SSH.defaultForwardedPorts is ignored in workspace configuration

### ISSUE THREAD

**Issue by qwabra** [microsoft/vscode-remote-release#3406]
- VSCode Version: 1.47.2
- Local OS Version:
- Remote OS Version: windows wsl 1 debian
- Remote Extension/Connection Type: SSH

# workspace configuration

.vscode/settings.json
```jsonp
{
    "typescript.tsdk": "node_modules/typescript/lib",
    "remote.SSH.defaultForwardedPorts": [
        {
            "name": "servePort",
            "localPort": 3000,
            "remotePort": 3000,
        },
        {
            "name": "HMRPort",
            "localPort": 10000,
            "remotePort": 10000,
        },
    ]
}
```

----

user/remote settings is ok

----

and ..., look like autosuggest is gone (?in forwarded ports tab after yarn dev)

**Comment by Chuxel** [comment]
@roblourens @sandy081 Should workspace settings be respected in this case?

**Comment by sandy081** [comment]
Yes, unless if the setting is declared otherwise.

**Comment by roblourens** [comment]
There is actually no way currently for an extension to know when vscode has connected after the resolver runs. I could cheat and use some API that I know will block until we are connected? But really we should have a nice way to wait for this.

**Comment by roblourens** [comment]
I have a fix, it will be in the next nightly.

**Comment by connor4312** [comment]
This isn't working for me, am I doing something wrong? https://memes.peet.io/img/20-08-2b6ca1c8-c105-4c16-842b-21060dd21399.webm

**Comment by roblourens** [comment]
Do you have any SSH settings set? Also can you share the log from the Remote-SSH output channel?

**Comment by connor4312** [comment]
Only `remote.SSH.remotePlatform`

I actually tried again just now and it did forward them correctly. But then I reloaded VS Code, and it didn't on the next reconnection. Refreshing a few more times, I didn't get my ports forwarded again. I tried killing the remote server and reconnecting, but that didn't seem to make a difference.

Log from when it failed:

```
[22:28:05.624] Log Level: 2
[22:28:05.626] remote-ssh-nightly@2020.8.3780
[22:28:05.626] win32 x64
[22:28:05.627] SSH Resolver called for "ssh-remote+home-server", attempt 1
[22:28:05.627] SSH Resolver called for host: home-server
[22:28:05.627] Setting up SSH remote "home-server"
[22:28:05.667] Using commit id "cfbd1999769f4f08dce29629fb92fdc0fac53829" and quality "insider" for server
[22:28:05.670] Install and start server if needed
[22:28:05.675] Checking ssh with "ssh -V"
[22:28:05.713] > OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5

[22:28:05.718] Running script with connection command: ssh -T -D 10620 home-server bash
[22:28:05.722] Terminal shell path: C:\WINDOWS\System32\cmd.exe
[22:28:05.909] > Enter passphrase for key 'C:\Users\Connor/.ssh/id_rsa':]0;C:\WINDOWS\System32\cmd.exe
[22:28:05.910] Got some output, clearing connection timeout
[22:28:05.911] Detected passphrase message
[22:28:08.696] Got passphrase response
[22:28:08.696] "install" wrote data to terminal: "*****************"
[22:28:08.710] > 
[22:28:08.910] > 0e6c8589d1fe: running
[22:28:08.966] > Acquiring lock on /home/connor/.vscode-server-insiders/bin/cfbd1999769f4f08dce29
> 629fb92fdc0fac53829/vscode-remote-lock.connor.cfbd1999769f4f08dce29629fb92fdc0fa
> c53829
[22:28:08.987] > Found existing installation at /home/connor/.vscode-server-insiders/bin/cfbd1999
> 769f4f08dce29629fb92fdc0fac53829...
[22:28:09.002] > Found running server...
> 
> *
> * Reminder: You may only use this software with Visual Studio family products, 
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
> 
[22:28:09.016] > Checking server status on port 40129 with wget
> 0e6c8589d1fe: start
> sshAuthSock====
> listeningOn==40129==
> osReleaseId==ubuntu==
> arch==x86_64==
> webUiAccessToken====
> tmpDir==/run/user/1000==
> platform==linux==
> 0e6c8589d1fe: end
[22:28:09.017] Received install output: 
sshAuthSock====
listeningOn==40129==
osReleaseId==ubuntu==
arch==x86_64==
webUiAccessToken====
tmpDir==/run/user/1000==
platform==linux==

[22:28:09.018] Remote server is listening on 40129
[22:28:09.018] Parsed server configuration: {"remoteListeningOn":{"port":40129},"osReleaseId":"ubuntu","arch":"x86_64","webUiAccessToken":"","sshAuthSock":"","tmpDir":"/run/user/1000","platform":"linux"}
[22:28:09.020] Starting forwarding server. localPort 10627 -> socksPort 10620 -> remotePort 40129
[22:28:09.022] Forwarding server listening on 10627
[22:28:09.023] Waiting for ssh tunnel to be ready
[22:28:09.025] [Forwarding server 10627] Got connection 0
[22:28:09.027] Tunneled 40129 to local port 10627
[22:28:09.027] Resolved "ssh-remote+home-server" to "127.0.0.1:10627"
[22:28:09.047] ------




[22:28:09.056] [Forwarding server 10627] Got connection 1
[22:28:09.057] [Forwarding server 10627] Got connection 2

```

**Comment by roblourens** [comment]
[remote-ssh-nightly-0.53.0.zip](https://github.com/microsoft/vscode-remote-release/files/5043134/remote-ssh-nightly-0.53.0.zip)

I tried this on mac and windows a bunch of times, and it never failed. I'm not sure what's going on, but there are a couple possibilities. Can you try this build with some extra logging?

**Comment by connor4312** [comment]
Here it is:

```
[13:46:43.305] Log Level: 2
[13:46:43.308] remote-ssh-nightly@0.53.0
[13:46:43.308] win32 x64
[13:46:43.311] SSH Resolver called for "ssh-remote+home-server", attempt 1
[13:46:43.311] SSH Resolver called for host: home-server
[13:46:43.311] Setting up SSH remote "home-server"
[13:46:43.354] Using commit id "2277c8e2a3e1cc630a6397301ba54a1dccd8a60d" and quality "insider" for server
[13:46:43.356] Install and start server if needed
[13:46:43.360] Checking ssh with "ssh -V"
[13:46:43.395] > OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5

[13:46:43.399] Running script with connection command: ssh -T -D 7531 home-server bash
[13:46:43.401] Terminal shell path: C:\WINDOWS\System32\cmd.exe
[13:46:43.600] > Enter passphrase for key 'C:\Users\Connor/.ssh/id_rsa':]0;C:\WINDOWS\System32\cmd.exe
[13:46:43.600] Got some output, clearing connection timeout
[13:46:43.601] Detected passphrase message
[13:46:46.567] Got passphrase response
[13:46:46.567] "install" wrote data to terminal: "*****************"
[13:46:46.581] > 
[13:46:46.756] > d2524acd2dab: running
[13:46:46.800] > Acquiring lock on /home/connor/.vscode-server-insiders/bin/2277c8e2a3e1cc630a639
> 7301ba54a1dccd8a60d/vscode-remote-lock.connor.2277c8e2a3e1cc630a6397301ba54a1dcc
> d8a60d
[13:46:46.824] > Found existing installation at /home/connor/.vscode-server-insiders/bin/2277c8e2
> a3e1cc630a6397301ba54a1dccd8a60d...
[13:46:46.840] > Found running server...
> 
> *
> * Reminder: You may only use this software with Visual Studio family products, 
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
> 
[13:46:46.855] > Checking server status on port 46637 with wget
[13:46:46.868] > d2524acd2dab: start     
> sshAuthSock====
> listeningOn==46637==    
> osReleaseId==ubuntu==   
> arch==x86_64==
> webUiAccessToken====    
> tmpDir==/run/user/1000==
> platform==linux==       
> d2524acd2dab: end       
[13:46:46.869] Received install output:      
sshAuthSock====
listeningOn==46637==    
osReleaseId==ubuntu==   
arch==x86_64==
webUiAccessToken====    
tmpDir==/run/user/1000==
platform==linux==       

[13:46:46.870] Remote server is listening on 46637
[13:46:46.870] Parsed server configuration: {"remoteListeningOn":{"port":46637},"osReleaseId":"ubuntu","arch":"x86_64","webUiAccessToken":"","sshAuthSock":"","tmpDir":"/run/user/1000","platform":"linux"}
[13:46:46.871] Starting forwarding server. localPort 7535 -> socksPort 7531 -> remotePort 46637
[13:46:46.872] Forwarding server listening on 7535
[13:46:46.872] Waiting for ssh tunnel to be ready
[13:46:46.874] Tunneled 46637 to local port 7535
[13:46:46.874] Resolved "ssh-remote+home-server" to "127.0.0.1:7535"
[13:46:46.875] [Forwarding server 7535] Got connection 0
[13:46:46.890] start guessRemoteCloud
[13:46:46.891] ------




[13:46:46.905] [Forwarding server 7535] Got connection 1
[13:46:46.906] [Forwarding server 7535] Got connection 2
[13:46:57.849] done guessRemoteCloud
[13:46:57.849] Will forward ports: 3000,10000
```

Something I notice is that `done guessRemoteCloud` took an extra 9 seconds to finish, and port forwarding didn't happen until then. It seems like, sshing to a kubuntu machine on my local network, it consistently takes about 9-11 seconds before the guess completes and port forwarding works after  that. Before I was just impatient and expected the forwarding to be instant.
